batch2/batch/worker.py

```diff
@@ -36,213 +38,210 @@
+        self.container_status = None
+        self.log = None
+
+    def container_config(self):
         config = {
             "AttachStdin": False,
             "AttachStdout": False,
             "AttachStderr": False,
             "Tty": False,
             'OpenStdin': False,
             'Cmd': self.spec['command'],
-            'Image': self.spec['image'],
+            'Image': self.image,
             'HostConfig': {'CpuPeriod': 100000,
-                           'CpuQuota': self.cores_mcpu * 100}
+                           'CpuQuota': self.cpu_in_mcpu * 100}
         }
 
         volume_mounts = self.spec.get('volume_mounts')
         if volume_mounts:
             config['HostConfig']['Binds'] = volume_mounts
 
-        n_tries = 1
-        error = None
-        image = config['Image']
-        while n_tries <= 3:
-            try:
-                self._container = await docker.containers.create(config, name=self.id)
-                self._container = await docker.containers.get(self._container._id)
-                return True
-            except DockerError as create_error:
-                log.info(f'Attempt {n_tries}: caught error while creating container {self.id}: {create_error.message}')
-                if create_error.status == 404:
-                    try:
-                        log.info(f'pulling image {image} for container {self.id}')
-                        await docker.pull(image)
-                    except DockerError as pull_error:
-                        log.info(f'caught error pulling image {image} for container {self.id}: {pull_error.status} {pull_error.message}')
-                        error = ImagePullBackOff(msg=pull_error.message)
-                else:
-                    error = Error(reason='Unknown', msg=create_error.message)
+        return config
 
-            await asyncio.sleep(1)
-            n_tries += 1
+    def step(self, state=None, name=None):
+        return ContainerStepManager(self, state, name)
 
-        self.error = error
-        return False
-
-    async def run(self):
-        assert self.error is None
-        log.info(f'running container {self.id}')
-
-        n_tries = 1
-        error = None
-
-        while n_tries <= 5:
-            try:
-                await self._container.start()
-                await self._container.wait()
-                error = None
-                break
-            except DockerError as err:
-                log.info(f'Attempt {n_tries}: caught error while starting container {self.id}: {err.message}, retrying')
-                error = RunContainerError(err.message)
-
-            await asyncio.sleep(1)
-            n_tries += 1
+    async def get_container_status(self):
+        c = await docker_call_retry(self.container.show)
+        log.info(f'container {self.pod.name}/{self.name}" container info {c}')
+        cstate = c['State']
+        status = {
+            'state': cstate['Status'],
+            'started_at': cstate['StartedAt'],
```

Comment:
Thanks.  I'm not using those for anything, I'm just passing them along because the previous interface did.  For durations I'm timing things directly.  It's possible `time.time` has the same problem; we should probably be using `time.monotonic` instead.  I'll put that on the todo to investigate.